File: QA-project/backend/src/db/schema/defect-lifecycle.ts
import {
  pgTable,
  serial,
  integer,
  varchar,
  timestamp,
} from "drizzle-orm/pg-core";
import { relations } from "drizzle-orm";
import { defects } from "./defects";

export const defectLifecycle = pgTable("defect_lifecycle", {
  lifecycleId: serial("lifecycle_id").primaryKey(),
  defectId: integer("defect_id")
    .notNull()
    .references(() => defects.defectId, { onDelete: "cascade" }),
  fromStatus: varchar("from_status", { length: 50 }),
  toStatus: varchar("to_status", { length: 50 }),
  changedAt: timestamp("changed_at").defaultNow().notNull(),
  reopenCount: integer("reopen_count").default(0),
  resolutionDays: integer("resolution_days"),
});

export const defectLifecycleRelations = relations(defectLifecycle, ({ one }) => ({
  defect: one(defects, {
    fields: [defectLifecycle.defectId],
    references: [defects.defectId],
  }),
}));

File: QA-project/backend/app/models/defect_lifecycle.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship
from app.database import Base


class DefectLifecycle(Base):
    __tablename__ = "defect_lifecycle"

    lifecycle_id = Column(Integer, primary_key=True, autoincrement=True)
    defect_id = Column(Integer, ForeignKey("defects.defect_id", ondelete="CASCADE"), nullable=False)
    from_status = Column(String(50))
    to_status = Column(String(50))
    changed_at = Column(DateTime, server_default=func.now(), nullable=False)
    reopen_count = Column(Integer, server_default="0")
    resolution_days = Column(Integer)

    # Relationships
    defect = relationship("Defect", back_populates="lifecycle")
